---
title: 'Unipept'
subtitle: 'Computational Exploration of Metaproteome Data'
author: 'Bart Mesuere'

lang: "en"

link-citations: true

titleDelim: ""
figureTemplate: "<span class='figure-title'>$$figureTitle$$&nbsp;$$i$$</span> $$t$$"
figPrefix:
  - "figure"
  - "figures"
tableTemplate: "<span class='table-title'>$$tableTitle$$&nbsp;$$i$$</span> $$t$$"
tblPrefix:
  - "table"
  - "tables"
secPrefix:
  - "section"
  - "sections"
listingTemplate: "<span class='listing-caption'><span class='listing-title'>$$listingTitle$$&nbsp;$$i$$</span> $$t$$</span>"
lstPrefix:
  - "listing"
  - "listings"
---
# Summary {- .front-matter data-running-title='Summary'}

Over the period of evolution, we human beings have co-evolved an intricate symbiosis with microorganisms that inhabit our gastrointestinal tract. These microorganisms are responsible for maintaining a healthy gut environment, they aid in digestion of our food and our immune system and they guard us against invading pathogens. In addition, some diseases, such as Crohn's disease, are somehow correlated to the composition of the gut microbiota. Although we are dependent on microorganisms for normal gut functioning, much remains to be learned about microbial processes in the gut that are carried out by this huge community of largely unexplored microbial cells that can amount to numbers as great as 10<sup>11</sup> per gram of faeces.

Recently, we have been aided by the development of molecular tools that enable us to determine the composition of microorganisms inhabiting the intestine without having to cultivate them. In addition to the increasing amounts of information about the identities of microorganisms in the gut from our own studies and others, there have been a limited number of studies of the functional genes in the entire gut microbial metagenome, using sequencing based metagenomics approaches.

A next step is to determine what genes are actually expressed and the function of the gut microbiota in different states of health and disease. Shotgun proteomics (@Fig:ch1fig3) is one approach that can be used to determine what proteins were expressed in an environmental sample. As of today, this technique is still in its infancy, but given the rapid technological developments and based on the results of the first analyses, we can nevertheless consider this to be a very promising technique.

Microorganisms drive most of the chemical transformations crucial to sustaining life on Earth. Their ability to inhabit almost every environmental niche proves they possess an incredible diversity of physiological capabilities. However, little is known about the majority of the millions of microbial species that are predicted to exists, given that we are able to grow only an estimated 1% of these organisms under lab conditions. The emerging disciplines of metagenomics and metaproteomics take advantage of the current generation of sequencing technologies to recover genetic material and active proteins directly from environmental samples. These new approaches provide us with a “new kind of microscope” that is revolutionizing our understanding of the diversity and ecology of environmental communities. However, the computational and statistical tools to analyze metagenomics and metaproteomics data are clearly lagging behind the developments in sequencing technology. To counter this, researchers at Ghent University are developing an online web portal called Unipept that combines advanced algorithms, novel statistical methods and interactive visualizations for the analysis of metagenomics and metaproteomics data sets. It will equip the new microscope with more powerful lenses, enabling researchers to better zoom in on who is living in complex environmental communities, what they are doing there and how they are doing it. To validate that Unipept has important applications across the fields of environmental, clinical and evolutionary microbiology, it is used to study shifts in the gut microbiota of patients suffering from cystic fibrosis (also known as mucovisidosis), the most common genetic disease in people from Caucasian origin that mostly affects the lungs and for which no cure exists today.

# Samenvatting {- .front-matter data-running-title='Samenvatting'}

# List of Publications {- .front-matter data-running-title='List of Publications'}

Tsilia, Varvara, Bart Devreese, Ilse Baenst De, **Bart Mesuere**, Andreja Rajkovic, Mieke Uyttendaele, Tom Van De Wiele, and Marc Heyndrickx. 2012. “Application of MALDI-TOF mass spectrometry for the detection of enterotoxins produced by pathogenic strains of the Bacillus cereus group.” *Analytical and Bioanalytical Chemistry* 404 (6-7): 1691–1702.<span class="hidden">@Tsilia2012</span>

**Mesuere, Bart**, Bart Devreese, Griet Debyser, Maarten Aerts, Peter Vandamme, and Peter Dawyndt. 2012. “Unipept: Tryptic peptide-based biodiversity analysis of metaproteome samples.” *Journal of Proteome Research* 11 (12): 5773–80. <span class="hidden">@Mesuere2012</span>

**Mesuere, Bart**, Griet Debyser, Maarten Aerts, Bart Devreese, Peter Vandamme, and Peter Dawyndt. 2015. “The Unipept metaproteomics analysis pipeline.” *Proteomics* 15 (8): 1437–42. <span class="hidden">@Mesuere2015</span>

Debyser, Griet, **Bart Mesuere**, Lieven Clement, Jens Van de Weygaert, Pieter Van Hecke, Gwen Duytschaever, Maarten Aerts, Peter Dawyndt, Kris De Boeck, Peter Vandamme, and Bart Devreese. 2015. “Faecal proteomics: A tool to investigate dysbiosis and inflammation in patients with cystic fibrosis.” *Journal of Cystic Fibrosis*. <span class="hidden">@debyser2015</span>

**Mesuere, Bart**, Toon Willems, Felix Van der Jeugt, Bart Devreese, Peter Vandamme, and Peter Dawyndt. 2016. “Unipept web services for metaproteomics analysis.” *Bioinformatics*. <span class="hidden">@Mesuere2016</span>

**Mesuere, Bart**, Felix Van der Jeugt, Bart Devreese, Peter Vandamme, and Peter Dawyndt. 2016. “The unique peptidome: taxon-specific tryptic peptides as biomarkers for targeted metaproteomics.” *Proteomics*. (submitted)
