# Peptidome analysis {.chapter data-running-title='4. Peptidome analysis'}

This chapter describes the Unique Peptide Finder, a tool for finding taxon-specific peptides that can be used as biomarkers for targeted proteomics. Together with the Peptidome Similarity feature, a tool for comparing the similarity of proteomes, the Unique Peptide Finder forms the Peptidome Analysis part of the Unipept web application.

## The unique peptidome: taxon-specific tryptic peptides as biomarkers&nbsp;for targeted metaproteomics {data-running-title='The unique peptidome' #sec:unique-peptidome}

<p class='aside'>This section contains a verbatim copy of the research article by @Mesuere2016a as submitted to Proteomics.</p>

**Abstract** &mdash; The unique peptide finder ([http://&#8203;unipept&#8203;.ugent&#8203;.be/&#8203;peptidefinder](http://unipept.ugent.be/peptidefinder)) is an interactive web application to quickly hunt for tryptic peptides that are unique to a particular species, genus or any other taxon. Biodiversity within the target taxon is represented by a set of proteomes selected from a monthly updated list of complete and non-redundant UniProt proteomes, supplemented with proprietary proteomes loaded into persistent local browser storage. The software computes and visualizes pan and core peptidomes as unions and intersections of tryptic peptides occurring in the selected proteomes. In addition, it also computes and displays unique peptidomes as the set of all tryptic peptides that occur in all selected proteomes but not in any UniProt record not assigned to the target taxon. As a result, the unique peptides can serve as robust biomarkers for the target taxon, e.g., in targeted metaproteomics studies. Computations are extremely fast since they are underpinned by the Unipept database, the Lowest Common Ancestor algorithm implemented in Unipept and modern web technologies that facilitate in-browser data storage and parallel processing.

### Introduction
The proteomics field is recently moving into more targeted approaches driven by key technologies such as Multiple Reaction Monitoring [@Pan2012;@Huttenhain2012]. The idea is to program triple quadrupole-like instruments to quantitatively determine predefined molecules by setting them to detect precursor ion masses and predefined MS/MS derived fragments. Target peptides can be selected according to a wide range of criteria such as their protein-specificity to screen for the presence/absence of certain proteins in proteomics studies or their taxon-specificity to screen for the presence/absence of certain taxa in metaproteomics studies. With protein profiling providing assays closer to activated functions, such metaproteome-wide association studies have the potential to become an important tool in environmental studies and modern medicine, and could answer major yet unmet environmental and clinical needs [@Juste2014].

In this paper we present the Unique Peptide Finder, an open source web application that enables the discovery of tryptic peptides that can be used as robust taxon-specific biomarkers. The software package is an integrated component of the Unipept ecosystem [@Mesuere2012;@Mesuere2015] and only takes a few milliseconds to compute all tryptic peptides that are not found in any known protein outside the target taxon and are shared by a sample of proteomes that represent the diversity within the target taxon. This set of tryptic peptides is called the unique peptidome and is visualized in a chart that also displays peptidome sizes and the core and pan peptidomes that give similar insights as the pan and core genomes [@Medini2005] but on a peptide level. Due to the focus on modern web technologies and extremely fast and parallel data processing the chart is highly interactive: including, removing, and rearranging proteomes results in almost instant recalculation of the unique peptidome. Both the data visualizations and the obtained sets of peptidomes can be easily exported for further processing.

### Web application
The backend of the Unique Peptide Finder is powered by a database that is rebuilt monthly using a completely automated pipeline. It contains all tryptic peptides extracted from the protein records in the UniProt Knowledgebase [@Wu2006], along with taxonomic information extracted from the NCBI Taxonomy database [@Wheeler2004]. These data sources are processed according to the procedures described in [@Mesuere2012], resulting in an index of tryptic peptides having a length between 5 and 50 amino acids and precomputed lowest common ancestors that represent their taxonomic distribution. The construction pipeline also preprocesses and integrates the list of all complete and non-redundant proteomes as published on the UniProt website. The resulting database provides quick retrieval of the set of tryptic peptides of all UniProt proteomes and the taxon-specificity of these tryptic peptides.

The web application itself was built using modern web technologies such as JavaScript (ES2015), HTML5, and CSS3. The D3.js library [@Bostock2011] was used to implement the interactive visualizations. Fetching sets of tryptic peptides for selected proteomes is the only step in the biomarker analysis that requires communication with the web server that runs the Ruby on Rails backend of Unipept. All other data processing is done client side using JavaScript workers. These workers run as a separate background processes to guarantee that the application remains responsive while the proteomes are being processed. Workers can request data from the database by doing asynchronous requests to the web server. Upon receiving such a request, the web server queries the database and sends the data back to the worker using the JSON (JavaScript Object Notation) format. The combination of these technologies results in a high performance, interactive web application with compelling visualizations without the need to install any software. The only requirement is a modern browser and an Internet connection.

The user interface of the Unique Peptide Finder is dominated by a large chart at the top of the page. Proteomes can be included, removed or rearranged using control elements underneath the chart or by direct interaction with the chart itself. The "Analyzed proteomes" at the right-hand side lists all proteomes currently included in the analysis. The "Proteome library" at the left-hand side lists all available proteomes: complete and non-redundant proteomes published by UniProt ("UniProt proteomes") and proteomes loaded into persistent local browser storage ("My proteomes"). The "UniProt proteomes" library currently contains over 10&thinsp;000 proteomes and is updated monthly with each new UniProt release. Additional proteomes can be added to the "My proteomes" library by loading multi-fasta files that each contain the protein sequences of a single proteome. Processed file contents are persistently stored in the browser so that they can be reused across intermittent sessions and are never stored on the Unipept server so that confidentiality of proprietary proteomes is never compromised. Proteomes are listed on pages of 50 hits and can be filtered on taxonomic rank, type strain status [@Federhen2014], reference status (as assigned by UniProt) and name ([@Fig:ch4fig1]). Plus-buttons can be clicked to include all or individual proteomes to the analysis.

<p style="display:none" class='pre-small-image'> </p> ![Screenshot of the proteome library with the UniProt proteomes filtered on *Eubacterium rectale*. The first hit is marked as a reference proteome by UniProt. The second hit is marked as a type strain by NCBI. Taxonomic breadcrumbs show class, order, genus, and species to which the proteomes were assigned. Clicking any of the breadcrumbs activates the corresponding taxon as a new taxonomic filter. Plus icons on the right can be clicked to include all or individual proteomes to the analysis.](images/ch4fig1.png){#fig:ch4fig1}

### Analysis
Each proteome included in the analysis is converted into its corresponding (tryptic) *peptidome*: the set of all (tryptic) peptides encoded in the proteome of the organism. The name of the first proteome included appears at the leftmost side of the chartâ€™s x-axis and a grey data point represents its peptidome size (the number of peptides in its peptidome). When a new proteome gets included in the analysis it appears at the rightmost side of the chart, with three additional data points representing sizes of the pan, core, and unique peptidomes of all proteomes that have been included in the analysis so far. The *pan peptidome* (blue data point) is the union of all these peptidomes and the *core peptidome* (orange data point) is their intersection. Peptides in the core peptidome occur in all proteomes included so far, but might also occur in proteomes of other taxa. To be useful as biomarkers for targeted metaproteomics, the core peptidome is further reduced by removing all peptides occurring in UniProt records that are not assigned to the target taxon. The target taxon is automatically computed as the lowest common taxonomic ancestor of all proteomes included in the analysis, using the Lowest Common Ancestor (LCA) algorithm implemented in Unipept [@Mesuere2012]. This reduced set is called the *unique peptidome* and is represented by a green data point.

All unions and intersections are computed solely from data already available in the browser. Determining unique peptidomes is more compute-intensive and requires knowledge about the taxon-specificity of tryptic peptides that is only available in the backend database. Their computations can be speeded up by carefully avoiding unnecessary computations and relying on precomputed lowest common taxonomic ancestors for all tryptic peptides in the Unipept database. Exact sizes of peptidomes are shown in a tooltip that appears when hovering over the data points in the chart ([@Fig:ch4fig2]). Clicking a data point displays a popup that provides options to remove the corresponding proteome from the analysis or download any of the four lists of peptides related to that proteome. This allows further data analysis or filtering of the unique peptidome (e.g., based on sequence length, peptide sequence, presence of amino acids that are prone to chemical modification, etc.) using specialized software for targeted (meta)proteomics such as Skyline [@MacLean2010].

![Chart displaying sizes of peptidomes (gray), pan peptidomes (blue), core peptidomes (orange) and unique peptidomes (green) for 32 *Acinetobacter baumannii* proteomes available in the UniProt proteomes library. The core peptidome of *A. baumannii* stabilizes gradually with 23&thinsp;579 tryptic peptides occurring in all proteomes included in the analysis. Of these common peptides, 1&thinsp;134 peptides in the unique peptidome are found to be specific for *A. baumannii*. The smaller peptidome size of *A. baumannii SDF* is no error in the application or reference database, but the result of a genome that is 20% reduced [@Fournier2006].](images/ch4fig2.png){#fig:ch4fig2}

Since pan, core, and unique peptidomes of a particular proteome are always computed based on the proteomes to its left in the chart, the order of proteomes included in the analysis matters. Proteomes can be easily rearranged by simply dragging them to the desired location in the chart or the "Analyzed proteomes" table. The latter also provides a couple of automatic rearrangement heuristics (e.g., alphabetically, on peptidome size, etc.). Removing proteomes is equally easy by dragging proteomes to the trash can on the chart, or by clicking the corresponding trash icon in the "Analyzed proteomes" table. Pan, core, and unique peptidomes that are affected after inclusions, removals or rearrangements are instantly recalculated and size updates are animated in the chart. The chart supports fullscreen mode that also provides all manipulations needed for biomarker analysis.

### Discussion
The rightmost unique peptidome in the chart theoretically represents robust taxon-specific biomarkers for metaproteomics analysis, since it contains tryptic peptides that occur in all the proteomes included in the analysis and not in a single UniProt record that is not assigned to the target taxon. In practice, however, sensitivity of the biomarkers depends on completeness of the included proteomes and how well they sample diversity within the target taxon. Their specificity depends on how well the global protein space has been sampled in the UniProt database. In addition, the analysis also depends on the correctness of the taxonomic annotations of the proteomes and UniProt entries. The value of the Unique Peptide Finder and some potential pitfalls are further discussed in the context of two case studies.

In a first case study we determined the unique peptidome of *Acinetobacter baumannii*. For this bacterial species, all 32 proteomes available in the "UniProt proteomes" library were included in the analysis. The proteome *A. baumannii 161/07* (UniProt Proteome UP000034216) with the largest peptidome size was chosen as the leftmost peptidome in the chart and the other proteomes were automatically sorted using the "optimize pan and core peptidome" heuristic ([@Fig:ch4fig2]). The resulting chart displays a steadily increasing pan peptidome (blue line), with an average of around 5&thinsp;000 proteome-specific peptides and an average peptidome size of around 70&thinsp;000. The core peptidome (orange line) levels off after a handful of proteomes, with a total of 23&thinsp;579 tryptic peptides shared by all included proteomes. Of these shared peptides 1&thinsp;134 do not occur in any species other than *A. baumannii*. As this unique peptidome (green line) contains peptides that are both common and unique to *A. baumannii*, it can be used to detect the occurrence of the species in metaproteome data sets.

The fact that biomarker analysis always relies on correct taxonomic assignments is illustrated by the misclassified proteome *A. baumannii 6411* (UniProt Proteome UP000031110) that was deliberately excluded from the above analysis. Using the Peptidome Clustering application (another tool from the Unipept ecosystem that is available online, unpublished) to cluster this proteome with available type strain proteomes of the genus *Acinetobacter*, shows that the proteome correctly classifies in the species *Acinetobacter nosocomialis*. As a result of this misclassification the Unique Peptide Finder discovers almost no unique peptides for the species for *A. nosocomialis*. Most of its actual unique peptides also occur in the proteome that was incorrectly classified as *A. baumannii*, so that the tool does not consider them to be specific to *A. nosocomialis*. The upside is that the Unique Peptide Finder in combination with the Peptidome Clustering tool allows to easily spot such misclassifications, after which they can be reported to curators of the underpinning information sources (in this case UniProt).

In a second case study we looked at potential biomarkers for *Bacillus anthracis*, a bacterial species that is hard to identify because of its relatedness with *Bacillus cereus*. For this species, the 3 proteomes available in the "UniProt proteomes" library were included in the analysis, together with proteomes of 24 other *B. anthracis* assemblies that were downloaded from the NCBI website (@Tbl:ch4tbl1) and loaded into the "My proteomes" library. The resulting chart (@Fig:ch4fig3) displays a core peptidome containing 50&thinsp;724 peptides, of which 878 are specific to *B. anthracis*. These unique peptides can be used to discern *B. anthracis* from *B. cereus* and any other species. When we revisit the 11 *B. anthracis* biomarkers proposed by @Chenau2014, we first observe that only 8 of them are true tryptic peptides. These 8 peptides are indeed found in *B. anthracis* proteomes and not in any *Bacillus cereus* proteome. However, 3 of these peptides are also found in non-*Bacillus cereus* proteomes: <span class="sequence">lvggvavik</span> in a range of other bacteria, <span class="sequence">ildqsadk</span> in several Firmicutes and <span class="sequence">vctitgr</span> in a *Turicibacter sanguinis* proteome. This can be easily verified using the Tryptic Peptide Analysis tool in Unipept. Only one of the 8 tryptic peptides (<span class="sequence">wllrpedpnyvlik</span>) was common to all 27 *B. anthracis* proteomes we analyzed in this case study.

### Conclusion
The Unique Peptide Finder is a user-friendly and high performance web application to find robust taxon-specific biomarkers for use in targeted metaproteomics studies. Biomarker analysis can combine a selection of UniProt proteomes and proprietary proteomes that are persistently stored in the browser. The tool is built using modern web technologies that provide parallel non-blocking computations, interactive visualizations and export functionality.

assembly name | assembly accession number
:------------ | :------------------------
Bacillus anthracis 52-G | GCA_000559005.1
Bacillus anthracis CZC5 | GCA_000534935.1
Bacillus anthracis (strain CDC 684 / NRRL 3495) | GCA_000021445.1
Bacillus anthracis str. Vollum | GCA_000742895.1
Bacillus anthracis str. Turkey32 | GCA_000833275.1
Bacillus anthracis str. Sterne | GCA_000832635.1
Bacillus anthracis str. V770-NP-1R | GCA_000832785.1
Bacillus anthracis 9080-G | GCA_000558985.1
Bacillus anthracis 8903-G | GCA_000558965.1
Bacillus anthracis (strain A0248) | GCA_000022865.1
Bacillus anthracis str. Carbosap | GCA_000732465.1
Bacillus anthracis str. A16 | GCA_000512835.1
Bacillus anthracis str. SVA11 | GCA_000583105.1
Bacillus anthracis Tsiankovskii-I | GCA_000181675.2
Bacillus anthracis str. A0174 | GCA_000182055.1
Bacillus anthracis str. A0193 | GCA_000181915.1
Bacillus anthracis str. A0389 | GCA_000219895.1
Bacillus anthracis str. A0488 | GCA_000181835.1
Bacillus anthracis str. A0465 | GCA_000181995.1
Bacillus anthracis str. A0442 | GCA_000181935.1
Bacillus anthracis str. BF1 | GCA_000295695.1
Bacillus anthracis str. A16R | GCA_000512775.1
Bacillus anthracis str. UR-1 | GCA_000292565.1
Bacillus anthracis str. 95014 | GCA_000585275.1

: Overview of the *Bacillus anthracis* assemblies used to analyze the *B. anthracis* peptidome. The order of the assemblies corresponds to the order of the proteomes in [@Fig:ch4fig3] {#tbl:ch4tbl1}

![Chart displaying sizes of peptidomes (gray), core peptidomes (orange) and unique peptidomes (green) for 3 *B. anthracis* proteomes available in the "UniProt proteomes" library and 24 *B. anthracis* proteomes downloaded from the NCBI website. These 27 proteomes have 50&thinsp;724 tryptic peptides in common (core peptidome), of which 878 are specific to *B. anthracis* (unique peptidome). Sizes of the pan peptidomes were hidden in the visualization to improve legibility of the chart.](images/ch4fig3.png){#fig:ch4fig3}

## Peptidome clustering {data-running-title='Peptidome clustering' #sec:peptidome-clustering}

Using the foundations of the Unique Peptide Finder (@Sec:unique-peptidome), the Peptidome Clustering feature allows users to calculate and visualize a similarity matrix for a set of proteomes. The proteomes, both the ones from UniProt as well as custom proteomes, can be selected using the same user interface as the Unique Peptide Finder. The similarity between two proteomes is determined by looking at the number of peptides they share in their respective peptidomes. Additionally, the hierarchical clustering method UPGMA is used to cluster the proteomes and generate a phylogenetic tree that is aligned with the similarity matrix (@Fig:ch4fig4).<span class='aside'>UPGMA stands for Unweighted Pair Group Method with Arithmetic Mean.</span> Each of the bifurcations of the phylogenetic tree contains a small circular button that enables users to swap the position of the two branches as well as the corresponding rows and columns of the similarity matrix. As with all Unipept visualizations, the image can be easily exported to both png and svg formats. The source data of the similarity matrix can be exported as a csv file containing all similarity values and the phylogenetic tree can be exported in the Newick tree format.

![Peptidome clustering of all 15 *Bacillus anthracis* proteomes and the 4 *Bacillus cereus* reference proteomes of UniProt using the "union" similarity measure. The species *B. anthracis* and *B. cereus* are considered very related and hard to distinguish, but the Unipept Peptidome Clustering shows a clear distinction between them.](images/ch4fig4.png){#fig:ch4fig4}

##### Similarity measures
The similarity measure used by the Peptidome Clustering is based on the ratio of peptides that two proteomes have in common. The nominator of this fraction is thus the number of peptides that both have in common, or the size of the intersection of the two sets of peptides. For the value of the denominator, the application offers five options: "Union" is the default option and uses the size of the union of the two sets. This is the recommended option for comparing multiple strains of a single species or closely related species and will always result in the lowest similarity values of the available options. The "Minimum" setting uses the minimum of the size of the two sets and will result in the highest similarity of the five options. This setting is recommended for comparing separate species. The three other options, "Maximum", "Average", and "Ochiai" will result in similarity values that fall between the previous two and will respectively use the maximum, average or square root of the product of the size of the two peptide sets.

##### Performance
All similarity calculations are done client-side in the web browser of the user, so the entire peptidome (i.e., the set of tryptic peptides) of each of the proteomes in the analysis needs to be downloaded. Downloading the actual sequences would take too much time, so another solution was needed. Since we only want to know if a peptide is present in two sets, the actual sequences are not needed, only something to uniquely identify them. These identifiers were found in the form of the primary keys used in the underlying Unipept database. Using these idâ€™s has two advantages: they are integers and thus smaller than the sequences themselves and they are faster to retrieve from the database. Using integers also allowed us to perform an additional encoding step know as delta encoding to further reduce the size.<span class="aside">Delta encoding is explained in more detail in @Sec:unipept-3-1.</span> Delta encoding transmits the differences between successive values instead of the values themselves. Because our list of integers is sorted, this encoding step results in smaller values and thus a smaller download size. Finally, all data is compressed by the web server using gzip. Gzip is a compression algorithm that is widely supported and built-in into most browsers. The combination of these processing steps allowed us to reduce the download size significantly. For a typical *Escherichia coli* genome, the size was reduced from 1.4 MB for the original peptide sequences to only 63 KB, a spectacular reduction of over 20 times.

To efficiently calculate similarities, we need to be able to quickly compute the union and intersection of sets. Unfortunately, JavaScript has no reliable implementation of the set data type. We had to find an alternative, but sadly none of the set implementations we tried met our performance requirements. Eventually, we used a different strategy and went with simple arrays of the integer identifiers described in the previous paragraph. When these arrays contain sorted integers, it is possible to efficiently calculate the union or intersection of them by simply synchronously iterating over them a single time. Because we donâ€™t need the union and intersection itself but only their sizes, we can deduce both sizes by only calculating one of the two and combining it with the already known sizes of the peptidomes. Exploiting the symmetry of the similarity matrix further reduces the computation time by half to under a millisecond per similarity. Calculating the entire similarity matrix for 33 *Acinetobacter baumannii* genomes takes around 180 milliseconds.
