# The Unique Peptide Finder {.chapter data-running-title='The Unique Peptide Finder'}

This chapter describes the Unique Peptide Finder, a tool for finding taxon-specific peptides that can be used as biomarkers for targeted proteomics. Together with the Peptidome Similarity feature, the Unique Peptide Finder forms the Peptidome Analysis part of the Unipept web application.

## The unique peptidome: finding taxon-specific biomarkers for targeted proteomics {data-running-title='The Unique Peptidome'}

**Abstract** — The unique peptide finder ([http://&#8203;unipept&#8203;.ugent&#8203;.be/&#8203;peptidefinder](http://unipept.ugent.be/peptidefinder)) is an interactive web application that lets users find unique tryptic peptides for a given species, genus or any other taxon in mere seconds. The application is part of the Unipept toolbox and is powered by the Unipept peptide database. The proteome data is retrieved from the list of complete and non-redundant UniProt proteomes, supplemented with any user-added proteomes. The unique peptide finder visualizes the union and intersection of a list of proteomes, similar to the pan and core genome but on a peptide level. It then calculates the unique peptides by taking all tryptic peptides occurring in all selected proteomes but not in any other UniProt record that is assigned to another taxon. Computations are extremely fast, since they are based on the Lowest Common Ancestor algorithm that has been implemented in Unipept. These unique peptides can be used as marker peptides for targeted proteomic experiments.

### Introduction
Recently, the proteomics field is undergoing a transition to more targeted approaches. A key technology is Multiple Reaction Monitoring (MRM; @Pan2012;@Huttenhain2012). This is a strategy in which triple quadrupole-like instruments are programmed to quantitatively determine a predefined molecule by setting them to detect the precursor ion mass and predefined MS/MS derived fragments (called a transition). Target peptides are then selected using a wide range of criteria: their specificity to be used as a biomarker for the presence/absence of a protein and, in a metaproteomics context, for the presence/absence of the species from which this protein is derived. With protein profiling providing assays closer to activated functions, such metaproteome-wide association studies have the potential to become an important tool in environmental studies and modern medicine, and could answer major yet unmet environmental and clinical needs [@Juste2014].

In this article we present an open source tool that allows to hunt for potential biomarkers in the form of tryptic peptides. The web application is part of the Unipept toolbox and lets users explore the common and distinct tryptic peptides of proteomes in a graph that is similar to the pan and core genome [@Medini2005], but on a peptide level. The set of core peptides that only occurs in the selected taxon and not in any other can easily be exported and used as biomarker candidates.

### Web Application
The Unique Peptide Finder is powered by a relational database containing all protein information from the UniProt Knowledgebase [@Wu2006] along with the taxonomic information from the NCBI Taxonomy database [@Wheeler2004]. These data sources are parsed according to the procedure described in @Mesuere2012 and @Mesuere2015 which results in an index of all tryptic peptides with a length between 5 and 50 amino acids. The proteome information is derived from the list of complete, non-redundant proteomes as published on the UniProt website. The resulting database allows us to quickly retrieve the set of tryptic peptides associated with a proteome.

The web application itself was built using modern web technologies such as JavaScript (ES2015), HTML5 and CSS3. The D3.js library [@Bostock2011] was used to create the graph and interactions. After receiving the proteome content from the web sever, all calculations are done client side by a JavaScript worker. This worker runs as a separate background process, which means that the web application stays responsive while data is being processed. The worker can request data from the database by doing an asynchronous request to the web server. The web server, running the Ruby on Rails Unipept application, queries the database and sends the data back to the worker using the JSON (JavaScript Object Notation) format. The combination of these technologies allowed us to create a high performance, interactive web application with compelling visualizations without the need to install any software. The only requirement is that the user has a modern browser and an internet connection.

The user interface of the Unique Peptide Finder is dominated by a large graph at the top of the page. Users can add proteomes to this graph by using the control elements underneath the graph. A table at the right hand side shows all proteomes currently added to the analysis. The left hand side shows the proteome library, a list of all available proteomes consisting of the the complete, non-redundant UniProt proteomes and a list of user-added proteomes. The UniProt proteome library contains over 10&thinsp;000 proteomes and is updated monthly with each new UniProt release. The proteomes are listed on pages of 50 results and can be filtered on taxonomic rank, type strain [@Federhen2014], reference proteome (as assigned by UniProt) and name ([@Fig:ch3fig1]). The resulting proteomes can easily be added to the analysis by clicking the plus-button. Users can also add proteomes that are not available in UniProt by adding a fasta file containing protein sequences. These files are persistently stored in the browser and not on the Unipept server which means that users can add their own proteomes without worrying about confidentiality and use them in intermittent browser sessions.

<p style="display:none" class='pre-small-image'> </p> ![Screenshot of the proteome library with a filter on *Eubacterium rectale*. The three results are the complete and non-redundant *E. rectale* proteomes that are available in UniProt. The first result is marked as a reference proteome by UniProt, the second result is a type strain. Under each proteome name, the class, order, genus and species of the organism are shown. Clicking on any of these links will adjust the filter for the corresponding taxon. The plus-buttons on the right can be used to add individual or all proteomes to the analysis.](images/ch3fig1.png){#fig:ch3fig1}

### Analysis
When a proteome gets added, the set of tryptic peptides belonging to that proteome is retrieved from the server if it is a UniProt proteome or from the browser storage, if it is a user-added proteome. Upon loading, the name of the proteome gets added to the rightmost side of the x-axis of the graph, and a gray dot representing the size of the set of peptides appears. When a second proteome gets loaded, two additional data points are calculated. A first one is the union of the newly loaded set of peptides with all previously loaded proteomes. We call this union the pan peptidome of the loaded proteomes. Analogously, the intersection with the previously loaded proteomes gets calculated. This intersection, which we call the core peptidome, is the set of tryptic peptides that all loaded proteomes have in common. The pan and core peptidome is similar to the pan and core genome but on a peptide level. The size of both the pan peptidome and the core peptidome are shown on the graph by a blue and orange dot respectively. The exact sizes of the sets are also shown in a tooltip when the user hovers over a proteome in the chart ([@Fig:ch3fig2]).

Since the union and intersection are always calculated based on the proteomes to the left of the current proteome, the order of the proteomes in the analysis matters. Users can easily rearrange the order of the proteomes by simply dragging them to the desired location in the graph or table. The affected pan and core peptidomes are instantly recalculated and the updated values are shown on the graph. The tool also has multiple options to automatically sort the proteomes (e.g. alphabetically, on peptidome size, ...). Removing proteomes is equally easy by dragging proteomes to the trash can on the graph, or by clicking the corresponding trash icon in the table.

The peptides of the core peptidome occur within each of the selected proteomes, but might also occur within other species. To be useful as a biomarker and for targeted proteomics, it would be convenient to filter this set to only retain peptides that occur within the species (or taxonomic group) of the chosen proteomes. To achieve this filtering, the Lowest Common Ancestor (LCA) algorithm of Unipept [@Mesuere2012] was used. The LCA algorithm searches for all UniProt entries where a given tryptic peptide is present, and uses the associated taxonomic lineage to determine the most specific taxonomic rank that all entries have in common. This set of peptides that only occurs within the selected taxon and within all selected proteomes is called the unique peptidome. The size of the unique peptidome is shown on the graph using green dots. By clicking any of the proteomes, a popup appears which enables users to download the corresponding peptidomes as a text file containing the list of peptides. The peptides of the unique peptidome can then be further analyzed and filtered for suitability (for example on peptide length, peptide sequence, presence of amino acids that are prone to chemical modification, etc.) in specialized software for targeted proteomics such as Skyline [@MacLean2010].

![Analysis of all complete, non-redundant *Acinetobacter baumannii* proteomes showing the sizes of the peptidome (gray), pan peptidome (blue), core peptidome (orange) and unique peptidome (green). The proteome with the biggest peptidome, *A. baumannii 161/07*, was chosen as the first (leftmost) peptidome in the graph, the other proteomes were automatically sorted using the “optimize pan and core peptidome” option. The core peptidome of *A. baumannii* is relatively stable, with 23&thinsp;579 tryptic peptides occurring in all 32 proteomes. Of these 23&thinsp;579 peptides, 1&thinsp;134 are only found in *A. baumannii* and can thus be used as biomarker for this species, for example in a targeted proteomics experiment. The smaller peptidome size of *A. baumannii SDF* is no error in the application or reference database, but the result of a 20% smaller genome [@Fournier2006].](images/ch3fig2.png){#fig:ch3fig2}

### Discussion
As an example, we will analyze the unique peptidome of *Acinetobacter baumannii* for which 32 complete and non-redundant proteomes are available in UniProt. After filtering the proteome library for *A. baumannii*, clicking on the blue plus-button adds all proteomes to the analysis resulting in @Fig:ch3fig2. On the resulting graph, we see that the pan peptidome (blue line) keeps rising steadily. This means that of the circa 70&thinsp;000 tryptic peptides of each peptidome, around 5&thinsp;000 are specific to that peptidome. The core peptidome (orange line) on the other hand levels off after the first few peptidomes. This means that there is a set of around 25&thinsp;000 tryptic peptides that are shared by all peptidomes. Of these 25&thinsp;000 peptides, more than 1&thinsp;100 peptides don’t occur in any other species and are unique to A. baumannii. We call this the unique peptidome of *A. baumannii* (green line). These tryptic peptides can be used as a starting point for biomarker discovery.

The quality of the returned results is heavily dependant on the annotations and taxonomic classifications of the publicly deposited material. An example of this is misclassified proteome “*Acinetobacter baumannii 6411*” (UniProt Proteome UP000031110) which is actually an *Acinetobacter nosocomialis* proteome. The result of this misclassification is that almost no unique peptides for A. nosocomialis can be found because the actual unique peptides also occur in one (incorrect) *A. baumannii* proteome making them non-unique. The upside is that the Unique Peptide Finder makes it easy to spot such errors after which they can be reported to the parties of interest.

As a second example, we look at potential biomarkers for *Bacillus anthracis*, a species that is hard to identify because of its relatedness with Bacillus cereus. We added the three non-redundant *B. anthracis* proteomes available in UniProt to an analysis, along with proteome data from 24 other *B. anthracis* assemblies downloaded from the NCBI website using the my proteomes feature. This resulted in the graph in @Fig:ch3fig3 which shows that of the 50&thinsp;724 tryptic peptides in the core peptidome, 878 are only found in *B. anthracis*. These peptides can potentially be used to discern *B. anthracis* from *B. cereus* (and any other organism). When we revisit the 11 biomarkers proposed by @Chenau2014, we find that 8 of them are tryptic peptides. All 8 of them can indeed be found in *B. anthracis* and not in any other Bacillus proteome. 3 peptides however can also be found in other non-*Bacillus anthracis* proteomes (<span class="sequence">lvggvavik</span> in several bacteria, <span class="sequence">ildqsadk</span> in several Firmicutes and <span class="sequence">vctitgr</span> in a single *Turicibacter sanguinis* proteome). Only one of the proposed tryptic peptides (<span class="sequence">wllrpedpnyvlik</span>) could be found in all of the 27 proteomes we analyzed.

![Analysis of the three complete, non-redundant *Bacillus anthracis* proteomes available in UniProt, along with 24 other *B. anthracis* proteomes downloaded from NCBI genomes that were added to the analysis through the my proteomes feature. The 27 proteomes have 50&thinsp;724 tryptic peptides in common (the core peptidome). Of these 50&thinsp;724 peptides, 878 are only found in *B. anthracis* and in no other UniProt records (the unique peptidome). The pan peptidome line was disabled in the visualization to improve the legibility of the graph.](images/ch3fig3.png){#fig:ch3fig3}

### Conclusion
The Unipept Unique Peptide Finder is a high performance web application that can be used to find biomarkers for use in targeted proteomics experiments. An analysis can be based on a combination of UniProt proteomes and user-added proteomes that are persistently stored in the user’s browser. The tool offers interactive visualizations and user-friendly export functionality to quickly find the set of unique tryptic peptides for any taxon.
